RFC.md

```diff
@@ -0,0 +1,466 @@
+		""" Combine embeds based on token mask """
+		...
+
+	def setup_caches(self, batch_size: int, dtype: torch.dtype) -> None:
+        """Setup key value caches for attention calculation.
+
+        Args:
+            batch_size (int): batch size for the caches.
+            dtype (torch.dtype): dtype for the caches.
+        """
+        self.decoder.setup_caches(batch_size, dtype)
+
+	def caches_are_enabled(self) -> bool:
+        """Check if the key value caches are setup."""
+        return self.decoder.caches_are_enabled()
+
+	def reset_caches(self):
+        """Reset the key value caches."""
+        self.decoder.reset_caches()
+
+	def forward(
+		self,
+        tokens: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+        **kwargs, # no need for encoder_mask
+    ) -> Tensor:
+
+    if 	encoder_input is not None:
+		encoder_embeds = self.encoder(**encoder_input)
+
+	tok_mask = tok != self.encoder_token
+	embeds = self.tok_embeddings(tokens[tok_mask])
+	embeds = self._merge_embeds(embeds, encoder_embeds, tok_mask)
+
+	output = self.decoder(tokens, mask, input_pos)
+	return output
+
+ ```
+
+
+### Fusion Layer
+This module's primary use case is for DeepFusion modules where you need to add new layers within the LLM decoder. Instead of adding additional layers directly, which would change layer counts and break checkpoint keys for the text model, FusionLayer fuses the new layer to an existing layer that it modifies. This fused layer is now treated as a single layer. The new parameters are returned with `fusion_params`. Two state dict hooks are used here so that the original layer doesn't have it's name modified in the `state_dict`.
+
+```python
+
+class FusionLayer(nn.Module):
+    """Fusion layer as introduced in `Flamingo: a Visual Language Model for
+    Few-Shot Learning <https://arxiv.org/abs/2204.14198>`_.
+
+    Deep Fusion model architectures combine pretrained encoder models with
+    pretrained language models by infusing the encoder outputs into the middle
+    layers of the LLM. This allow the language model to interpret the encoder
+    outputs as text, to "understand" different modalities that you can train an
+    encoder for. To enable the language model to adapt to the encoder outputs,
+    the FusionLayer inserts a new learnable layer between the decoder (language
+    model) layers to learn to combine the encoder outputs and decoder
+    activations. The fusion layer can be inserted before or after the decoder
+    layer, in Flamingo they insert before.
+
+    Args:
+        layer (nn.Module): original decoder layer
+        fusion_layer (nn.Module): new fusion layer
+        fusion_first (bool): boolean to insert fusion layer before or after
+	        the decoder layer.
+    """
+
+    def __init__(
+        self,
+        layer: nn.Module,
+        fusion_layer: nn.Module,
+        fusion_first: bool = True
```

Comment:
OK I do understand the motivation for keeping the state dict the same and that is a nice advantage. But this feels so awkward to me. Anyways I am just complaining, I know that neither of these approaches is gonna be perfect